### ISSUE TITLE
[BUG] Unexpected pair coeff mix behavior with hybrid/overlay

### ISSUE THREAD

**Issue by bendjensen** [lammps/lammps#2754]
**Summary**

I'm interested in combining `lj/mdf` and `coul/long` using `hybrid/overlay` since a `lj/mdf/coul/long` pair style does not exist. To demonstrate the unexpected behavior of hybrid/overlay the example uses `lj/cut` and `coul/cut` for simplicity. 

My expectation from reading the manual was that these inputs would all behave similarly:
```
pair_style lj/cut 9
pair_style lj/cut/coul/cut 9 9
pair_style hybrid/overlay lj/cut 9
pair_style hybrid/overlay lj/cut 9 coul/cut 9
```

Any of the first three lines will run as expected, but if the fourth is used it will produce the error:
> ERROR on proc 0: All pair coeffs are not set (src/pair_hybrid.cpp:659)

It's surprising to me, as a user, that LAMMPS can produce the mixed coeffs using standard mixing rules like geometric or arithmetic for `hybrid/overlay lj/cut` but cannot if `coul/cut` is added.

I'm not sure if this is a bug or intended behavior but it is surprising. I understand hybrid mixing can become complicated quickly but I was hoping that for this simple case there would be some option to allow standard mixing rules be applied to lj terms.

The current implementation also easily leads to incorrect/unintended results by missing mixed lj terms. Because coul/cut has no parameters if a user does: `pair_coeff * * coul/cut` the script will run with no errors but will produce incorrect lj values because mix terms will only be assigned coul/cut interactions. It is quite surprising that `hybrid/overlay lj/cut 9` and` hybrid/overlay lj/cut 9 coul/cut 9` can produce different van der Waals energies.

**LAMMPS Version and Platform**

LAMMPS (8 Apr 2021)

**Expected Behavior**

See above

**Actual Behavior**

See above

**Steps to Reproduce**

A simple script and datafile is attached, based off the the tiny_epoxy example in USER/reaction/tiny_epoxy, to reproduce the unexpected behavior.

**Further Information, Files, and Links**

[in.script.hybrid-overlay.txt](https://github.com/lammps/lammps/files/6435054/in.script.hybrid-overlay.txt)
[in.script.txt](https://github.com/lammps/lammps/files/6435056/in.script.txt)
[tiny_epoxy.data.txt](https://github.com/lammps/lammps/files/6435057/tiny_epoxy.data.txt)

**Comment by akohlmey** [comment]
@bendjensen this behavior may be surprising to you, but it is documented and required for other uses of the hybrid/overlay style. Please keep in mind that LAMMPS has no way to know  your intentions. And here it is unable to determine whether your intention is to apply lj/cut **only** to the pairs of types that you explicitly mentioned or all mixed terms between those as well. Thus **all** terms must be provided. For the regular hybrid style some mixing can be inferred if you designate a `pair_coeff n n lj/cut` and a `pair_coeff m m lj/cut` set of parameters and all 'n m' combinations are otherwise unassigned. But for hybrid/overlay this is ambiguous since you already have one other pair style (coul/cut or coul/long) assigned.

The simple solution is to set up the system without the hybrid and coulomb first. Use write_coeff to write out all mixed parameters and then insert and edit the terms you need into your input after switching to hybrid/overlay.

**Comment by bendjensen** [comment]
My reading of the documentation indicated that mixing is performed if the individual pair_style supported mixing for hybrid/overlay. 
https://lammps.sandia.gov/doc/pair_hybrid.html
> For atom type pairs I,J and I != J, if the sub-style assigned to I,I and J,J is the same, and if the sub-style allows for mixing, then the coefficients for I,J can be mixed. This means you do not have to specify a pair_coeff command for I,J since the I,J type pair will be assigned automatically to the sub-style defined for both I,I and J,J and its coefficients generated by the mixing rule used by that sub-style. For the hybrid/overlay style, there is an additional requirement that both the I,I and J,J pairs are assigned to a single sub-style. See the “pair_modify” command for details of mixing rules. See the See the doc page for the sub-style to see if allows for mixing.

Can the docs be modified the better clarify the LAMMPS behavior, that if more than one pair style is selected for hybrid/overlay no pair coefficients for I,J are generated, even if the individual pair styles support mixing individually. Something like this:
> For the hybrid/overlay style, there is an additional requirement that both the I,I and J,J pairs are assigned to a single sub-style. If this requirement is not met, no I,J coeffs will be generated, even if the sub-styles support mixing, and I,J pair coefficients must be explicitly defined.

I appreciate the tip on generating the mixed terms with write_coeff. I can run the simulations with the explicit definitions for I,J pairs. 

My motivation for submitting this is because in a common scenario like
```
pair_coeff * * coul/cut
pair_coeff 1 1 lj/cut 1 1
pair_coeff 2 2 lj/cut 2 2
```
LAMMPS is inferring the incorrect behavior, that no lj/cut interactions should be occur between mixed pairs, and runs without error or warning. I understand that LAMMPS is missing the information required to differentiate between this and a scenario where a user actually does want to exclude lj/cut from I,J interactions. Can this information be provided by pair_modify? Something like `pair_modify lj/cut mix geometric` vs `pair_modify lj/cut mix none`?

Side note: There is a typo in the docs, `pair_hybrid` should be `pair_style hybrid`

**Comment by akohlmey** [comment]
> My reading of the documentation indicated that mixing is performed if the individual pair_style supported mixing for hybrid/overlay.
> https://lammps.sandia.gov/doc/pair_hybrid.html
> 
> > For atom type pairs I,J and I != J, if the sub-style assigned to I,I and J,J is the same, and if the sub-style allows for mixing, then the coefficients for I,J can be mixed. This means you do not have to specify a pair_coeff command for I,J since the I,J type pair will be assigned automatically to the sub-style defined for both I,I and J,J and its coefficients generated by the mixing rule used by that sub-style. For the hybrid/overlay style, there is an additional requirement that both the I,I and J,J pairs are assigned to a single sub-style. See the “pair_modify” command for details of mixing rules. See the See the doc page for the sub-style to see if allows for mixing.
 
I think the "additional requirement" phrase says what LAMMPS does.

> Can the docs be modified the better clarify the LAMMPS behavior, that if more than one pair style is selected for hybrid/overlay no pair coefficients for I,J are generated, even if the individual pair styles support mixing individually. Something like this:
> 
> > For the hybrid/overlay style, there is an additional requirement that both the I,I and J,J pairs are assigned to a single sub-style. If this requirement is not met, no I,J coeffs will be generated, even if the sub-styles support mixing, and I,J pair coefficients must be explicitly defined.

This can be added, but it is effectively saying the same thing twice.

> My motivation for submitting this is because in a common scenario like

Please note, this may seem "common" to you, but there are many different simulation situations that people do.
If you worry this much, then the better approach is to make a copy of the pair style lj/cut/coul/long sources and modify it for lj/mdf so that you have the additional benefit of improved performance from having to walk the neighbor lists only once.
 
> ```
> pair_coeff * * coul/cut
> pair_coeff 1 1 lj/cut 1 1
> pair_coeff 2 2 lj/cut 2 2
> ```
> 
> LAMMPS is inferring the incorrect behavior, that no lj/cut interactions should be occur between mixed pairs, and runs without error or warning.

I disagree with use of "incorrect" here; it is "unintended behavior". LAMMPS is doing **exactly** what the input specifies. The mixing behavior is an *additional convenience*, mostly added for use with "bio" force fields that use it a lot, but mixing is only available for a small subset of the many pair styles in the first place. So from that perspective the "not mixing" behavior is the more common situation.

> I understand that LAMMPS is missing the information required to differentiate between this and a scenario where a user actually does want to exclude lj/cut from I,J interactions. Can this information be provided by pair_modify? Something like `pair_modify lj/cut mix geometric` vs `pair_modify lj/cut mix none`?

The code is already complex as it is as you can see from the complexity of the documentation for this aspect. Please also note that "pair_modify mix" determines the mixing formula and there is a default of geometric mixing. So to make changes there would require to change expected default behavior that would break a lot of people's inputs. The logic of **deciding** when to mix or not is in the hybrid pair style code itself. The problem of adding a command to change the behavior is to then do it also in a way that is not doing too much of it and having to check for that as well. I expect that this would at least double the amount of code required and would make debugging and maintaining it even more challenging than it already is.

Thus my suggestion of creating a new combined pair style is usually the simpler way out since the differences between the LJ variants are often very small.

> Side note: There is a typo in the docs, `pair_hybrid` should be `pair_style hybrid`

**Comment by bendjensen** [comment]
> > My reading of the documentation indicated that mixing is performed if the individual pair_style supported mixing for hybrid/overlay.
> > https://lammps.sandia.gov/doc/pair_hybrid.html
> > > For atom type pairs I,J and I != J, if the sub-style assigned to I,I and J,J is the same, and if the sub-style allows for mixing, then the coefficients for I,J can be mixed. This means you do not have to specify a pair_coeff command for I,J since the I,J type pair will be assigned automatically to the sub-style defined for both I,I and J,J and its coefficients generated by the mixing rule used by that sub-style. For the hybrid/overlay style, there is an additional requirement that both the I,I and J,J pairs are assigned to a single sub-style. See the “pair_modify” command for details of mixing rules. See the See the doc page for the sub-style to see if allows for mixing.
> 
> I think the "additional requirement" phrase says what LAMMPS does.

Yes, agreed that it does.
 
> > Can the docs be modified the better clarify the LAMMPS behavior, that if more than one pair style is selected for hybrid/overlay no pair coefficients for I,J are generated, even if the individual pair styles support mixing individually. Something like this:
> > > For the hybrid/overlay style, there is an additional requirement that both the I,I and J,J pairs are assigned to a single sub-style. If this requirement is not met, no I,J coeffs will be generated, even if the sub-styles support mixing, and I,J pair coefficients must be explicitly defined.
> 
> This can be added, but it is effectively saying the same thing twice.

The first two sentences in the paragraph are also effectively saying the same thing twice, which I think is a helpful way to communicate the LAMMPS behavior.

> > My motivation for submitting this is because in a common scenario like
> 
> Please note, this may seem "common" to you, but there are many different simulation situations that people do.
> If you worry this much, then the better approach is to make a copy of the pair style lj/cut/coul/long sources and modify it for lj/mdf so that you have the additional benefit of improved performance from having to walk the neighbor lists only once.
> 
> > ```
> > pair_coeff * * coul/cut
> > pair_coeff 1 1 lj/cut 1 1
> > pair_coeff 2 2 lj/cut 2 2
> > ```
> > 
> > 
> > LAMMPS is inferring the incorrect behavior, that no lj/cut interactions should be occur between mixed pairs, and runs without error or warning.
> 
> I disagree with use of "incorrect" here; it is "unintended behavior". LAMMPS is doing **exactly** what the input specifies. The mixing behavior is an _additional convenience_, mostly added for use with "bio" force fields that use it a lot, but mixing is only available for a small subset of the many pair styles in the first place. So from that perspective the "not mixing" behavior is the more common situation.

Yes, "unintended behavior" is a better term. And understood that mixing is pretty bio-specific and that a more common scenario for hybrid/overlay is mixing tersoff, LJ, EAM, coarse-grained, etc styles since lj/cut/coul/long and related styles cover most bio force fields. 

> 
> > I understand that LAMMPS is missing the information required to differentiate between this and a scenario where a user actually does want to exclude lj/cut from I,J interactions. Can this information be provided by pair_modify? Something like `pair_modify lj/cut mix geometric` vs `pair_modify lj/cut mix none`?
> 
> The code is already complex as it is as you can see from the complexity of the documentation for this aspect. Please also note that "pair_modify mix" determines the mixing formula and there is a default of geometric mixing. So to make changes there would require to change expected default behavior that would break a lot of people's inputs. The logic of **deciding** when to mix or not is in the hybrid pair style code itself. The problem of adding a command to change the behavior is to then do it also in a way that is not doing too much of it and having to check for that as well. I expect that this would at least double the amount of code required and would make debugging and maintaining it even more challenging than it already is.
> 

Understood. My intent in submitting this was as a service to the LAMMPs developers and users who may also run into this scenario in their own use-cases, and to verify this was the intended behavior. I do believe it is surprising that the lj/cut mixing service LAMMPS provides occurs in all cases with one exception of hybrid/overlay and multiple pair styles, but certainly understand the complications that are driving this. Understood that this may not be worth the extra complications it may introduce into the code to enable a mix style argument into hybrid/overlay.

> Thus my suggestion of creating a new combined pair style is usually the simpler way out since the differences between the LJ variants are often very small.

Thank you for the suggestion, I will probably take this approach for my particular situation. I just wanted to raise the issue in case it was indicative of a broader use-case.

> 
> > Side note: There is a typo in the docs, `pair_hybrid` should be `pair_style hybrid`

**Comment by bendjensen** [comment]
Thank you Axel!
